Dynamic Edge Profiling with Optimal Counter Placement (#134)
sampsyo/cs6120: content/blog/2019-11-13-DynamicEdgeProfiling/index.md

As the profiling logic and data are written in a runtime library disjoint from the LLVM pass, this information can be outputted however the user wants with minor edits to the library file. We chose to print it to standard out for simplicity. There is also the matter of correlating the block numbers assigned during the LLVM pass to the actual blocks, and this is done by printing the block-to-number matching for the source and destination blocks of each instrumented edge to standard out during the pass.

## Evaluation
### Correctness
To measure performance, we ran a subset of C benchmarks from the [LLVM test suite](https://llvm.org/docs/TestSuiteGuide.html). We collected the tabulated results and CFGs to verify our instrumentation was on an optimal number of edges. We then ran the algorithm described above to extrapolate the uninstrumented edges. We include one test here for visibility.
Consider the following CFGS for a program with two functions, main and testFunc.
 <!-- ![](https://i.imgur.com/h9Stzm2.png =300x) ![](https://i.imgur.com/ls7PM8e.png =300x)  -->
<img src="main.png"/>
<img src="testfunc.png"/>

> We run our instrumented program with input argument `5` and procure the table below.

Reviewer: ```suggestion
We run our instrumented program with input argument `5` and procure the table below.

```

Below are our performance results. The data has been normalized to the average runtime for each program's original code after LLVM `-O3`. This means that if a bar's height is 2, that stage's average runtime took 2x longer than the optimized but unprofiled code. 

<!-- ![](https://i.imgur.com/uICyHTe.png) -->
<img src="results.png" style="max-width: 100%" >

A next step to further improve performance would be to place counters on edges that are less likely to be executed, decreasing the number of dynamic instructions. This can be done by additional profiling and then creating a maximum spanning tree using the estimated edge weights instead of an arbitrary spanning tree.

## Challenges
One of the primary challenges of this project was correctly installing LLVM on each of our machines with the correct version and ```/include/``` directories. When planning our design, we utilized several online resources that varied quite a bit in terms of targeted LLVM version. Available source files change quite a bit, and therefore we realized we needed to be more aware of the version of LLVM these resources were using when reading them. 

> The other main challenge was figuring out how to work with LLVM. Part of this work encompassed determining the tools and information we had access to and what information we needed to collect manually. We found the most effective way to learn how to work with LLVM was from experimenting with examples. In particular, when implementing dynamic edge profiling, we found passes that implement other dynamic passes like dynamic instruction count to be especially helpful.

Reviewer: ```suggestion
The other main challenge was figuring out how to work with LLVM. Part of this work encompassed determining the tools and information we had access to and what information we needed to collect manually. We found the most effective way to learn how to work with LLVM was from experimenting with examples. In particular, when implementing dynamic edge profiling, we found passes that implement other dynamic passes like dynamic instruction count to be especially helpful.

The source code for the LLVM pass detailed above can be found on [GitHub](https://github.com/kavoor/llvm-pass-skeleton).
```

+++
title = "Dynamic Edge Profiling with Optimal Counter Placement"
[extra]
bio = """
  [Katy](https://github.com/kavoor) is a Senior interested in Compilers and Systems Programming. She enjoys playing soccer and exploring Ithaca in her free time.

>   [Henry](https://www.linkedin.com/in/liuhenry4428)

Reviewer 2: Henry is apparently just… Henry.

>   [Henry](https://www.linkedin.com/in/liuhenry4428)

Reviewer 3: ```suggestion
  [Henry](https://www.linkedin.com/in/liuhenry4428) is a Senior interested in Compilers and Systems Programming. He enjoys playing video games and exploring different foods on his free time. 
```

"""
+++

## Introduction

For this project, we wanted to implement an LLVM pass that dynamically profiled basic block edges. The naïve implementation would be to add a profiling mechanism to every edge in the program, but that comes with a significant amount of overhead. To minimize overhead, we strived to add profiling to the minimum number of edges necessary to obtain the same amount of profiling information. The key insight is that we can determine if an edge is traversed using profiling information from preceding and succeeding edges if the CFG structure ensures the program can only run through a simple path between those profiled edges.

## Algorithm
To select the edges to profile, we used Knuth's algorithm:

---
<!-- ![](https://i.imgur.com/NbavLnq.png) -->

> <img src="array.png" style="max-width: 100%" >

Reviewer 2: Yay; thanks for including the image file instead of hotlinking.

---

> Essentially, this algorithm first finds the (acyclic) minimum spanning tree of the edges in a function and then adds profiling instrumentation---code that logs profiling information---to all edges not in the spanning tree. Note that this algorithm uses the first block in a function as the root of the spanning tree, and that edges are treated as bidirectional when checking for cycles.

Reviewer 2: To state the obvious, minimum spanning trees, because they are trees, are acyclic by definition—so that parenthetical may confuse more than it clarifies.

> Essentially, this algorithm first finds the (acyclic) minimum spanning tree of the edges in a function and then adds profiling instrumentation---code that logs profiling information---to all edges not in the spanning tree. Note that this algorithm uses the first block in a function as the root of the spanning tree, and that edges are treated as bidirectional when checking for cycles.

Reviewer 3: ```suggestion
Essentially, this algorithm first finds the minimum spanning tree of the edges in a function and then adds profiling instrumentation---code that logs profiling information---to all edges not in the spanning tree. Note that this algorithm uses the first block in a function as the root of the spanning tree, and that edges are treated as bidirectional when checking for cycles.
```

This algorithm works because the acyclic nature of the spanning tree entails that there is exactly one path between any two instrumented edges (or the entry or exit of the function). Thus, logging the traversal of one instrumented edge implies traversal of the program through the path from the previously logged edge location to the current one. This also implies that this is the minimum number of instrumented edges because having one less instrumented edge would either create a path that could be traversed without logging any profiling information or create multiple paths that lack sufficient profiling information to deduce which path was taken. 

### Offline Edge Count Extrapolation Algorithm
After we run our program, we know the number of times our instrumented edges were traversed. In order to determine the number of times the remaining edges were produced, we need to do some post-processing of our CFG with our runtime results.

> The idea behind this algorithm is to iterate over the edges and fill in the edge count until convergence. The program will converge under the assumption that the edge profiling information supplied is sufficient to determine the rest of the edge counts. It is valid to fill in an edge count if there is a simple path between two nodes with known edge count. You can find detailed pseudocode and proofs for this offline algorithm in the [LLVM Publication](https://llvm.org/pubs/2010-04-NeustifterProfiling.pdf).

Reviewer 2: Instead of just saying "LLVM Publication," maybe refer to the paper by the title or the author? (Actually, it seems to be a dissertation, not a paper.)

> The idea behind this algorithm is to iterate over the edges and fill in the edge count until convergence. The program will converge under the assumption that the edge profiling information supplied is sufficient to determine the rest of the edge counts. It is valid to fill in an edge count if there is a simple path between two nodes with known edge count. You can find detailed pseudocode and proofs for this offline algorithm in the [LLVM Publication](https://llvm.org/pubs/2010-04-NeustifterProfiling.pdf).

Reviewer 3: ```suggestion
The idea behind this algorithm is to iterate over the edges and fill in the edge count until convergence. The program will converge under the assumption that the edge profiling information supplied is sufficient to determine the rest of the edge counts. It is valid to fill in an edge count if there is a simple path between two nodes with known edge count. You can find detailed pseudocode and proofs for this offline algorithm in this [dissertation](https://llvm.org/pubs/2010-04-NeustifterProfiling.pdf) by Andreas Neustifter.
```

## Implementation

> We wanted our instrumentation implementation to be as noninvasive as possible for the base code as pervasive changes could make the LLVM IR harder to reason about and could prevent future optimizations. To achieve that, we implemented the bulk of the profiling information logging in a separate runtime library file. This strategy allows us to reduce instrumentation to two function calls, one to `logsrc(num1)` in the edge's "source" block and one to `logdest(num2)` in the "destination" block, both of which reside in the library file. These functions calls take in a block-unique integer (`num1` and `num2` above) that is generated during the pass as arguments, meaning that we can identify an edge traversal with a pair of `logsrc(num1), logdest(num2)` function calls.

Reviewer 2: Just for fun, it would be cool to know how you determined that block-unique integer. Perhaps just the position in the function's block list?

> We wanted our instrumentation implementation to be as noninvasive as possible for the base code as pervasive changes could make the LLVM IR harder to reason about and could prevent future optimizations. To achieve that, we implemented the bulk of the profiling information logging in a separate runtime library file. This strategy allows us to reduce instrumentation to two function calls, one to `logsrc(num1)` in the edge's "source" block and one to `logdest(num2)` in the "destination" block, both of which reside in the library file. These functions calls take in a block-unique integer (`num1` and `num2` above) that is generated during the pass as arguments, meaning that we can identify an edge traversal with a pair of `logsrc(num1), logdest(num2)` function calls.

Reviewer 3: ```suggestion
We wanted our instrumentation implementation to be as noninvasive as possible for the base code as pervasive changes could make the LLVM IR harder to reason about and could prevent future optimizations. To achieve that, we implemented the bulk of the profiling information logging in a separate runtime library file. This strategy allows us to reduce instrumentation to two function calls, one to `logsrc(num1)` in the edge's "source" block and one to `logdest(num2)` in the "destination" block, both of which reside in the library file. These functions calls take in a block-unique integer (`num1` and `num2` above) that is generated during the pass as arguments, meaning that we can identify an edge traversal with a pair of `logsrc(num1), logdest(num2)` function calls. The unique integers were simply the order in which the blocks were iterated through in our pass.

As the profiling logic and data are written in a runtime library disjoint from the LLVM pass, this information can be outputted however the user wants with minor edits to the library file. We chose to print it to standard out for simplicity. There is also the matter of correlating the block numbers assigned during the LLVM pass to the actual blocks, and this is done by printing the block-to-number matching for the source and destination blocks of each instrumented edge to standard out during the pass.

## Evaluation
### Correctness

> To measure performance, we ran a subset of C benchmarks from the [LLVM test suite](https://llvm.org/docs/TestSuiteGuide.html). We collected the tabulated results and CFGs to verify our instrumentation was on an optimal number of edges. We then ran the algorithm described above to extrapolate the uninstrumented edges. We include one test here for visibility.

Reviewer 2: To clarify, that "verification" was by manual human inspection, right?

> To measure performance, we ran a subset of C benchmarks from the [LLVM test suite](https://llvm.org/docs/TestSuiteGuide.html). We collected the tabulated results and CFGs to verify our instrumentation was on an optimal number of edges. We then ran the algorithm described above to extrapolate the uninstrumented edges. We include one test here for visibility.

Reviewer: ```suggestion
To measure performance, we ran a subset of C benchmarks from the [LLVM test suite](https://llvm.org/docs/TestSuiteGuide.html). We collected the tabulated results and CFGs to manually verify our instrumentation was on an optimal number of edges with correct edge counts. We ran the algorithm described above to extrapolate the uninstrumented edges. We include one test here for visibility.
```

> Consider the following CFGS for a program with two functions, main and testFunc.

Reviewer 2: CFGs

> Consider the following CFGS for a program with two functions, main and testFunc.

Reviewer: ```suggestion
Consider the following CFGs for a program with two functions, main and testFunc.
```

 <!-- ![](https://i.imgur.com/h9Stzm2.png =300x) ![](https://i.imgur.com/ls7PM8e.png =300x)  -->
<img src="main.png"/>
<img src="testfunc.png"/>
We run our instrumented program with input argument `5` and procure the table below.

| Edge     | Count |
|----------|-------|
| [6->4]   | 5     |
| [9->10]  | 1     |

When we ran our pass, we also output the edges that were instrumented.
These edges were `[0->2]`, `[6->4]`, and `[9->10]`.

To verify these results, we consider our program. 
```
void testFunc(int n){
    if(n%2==1) printf("The number is odd\n");
}

int main(int num) {
    for(int i =0; i<5; i++){
        printf("Iteration %i\n", i);
    }
    if(num%2 == 0){
        printf("The number is even\n");
    }
    else{
        testFunc(num);
    }
    printf("%i\n", num + 2);
    return 0;
}
```

> In order to automatically verify these results, we implement a naive LLVM pass for edge profiling that instruments every edge. We then compare the results collected and extrapolated from optimal edge profiling with the results from running the naive implementation.

Reviewer 2: I notice that your earlier instance of "naïve" had an umlaut. Any reason these later occurrences were left out? :smiley:

> In order to automatically verify these results, we implement a naive LLVM pass for edge profiling that instruments every edge. We then compare the results collected and extrapolated from optimal edge profiling with the results from running the naive implementation.

Reviewer: ```suggestion
In order to automatically verify these results, we implement a naïve LLVM pass for edge profiling that instruments every edge. We then compare the results collected and extrapolated from optimal edge profiling with the results from running the naïve implementation.
```

> Consider the results from running the naive implementation.

Reviewer: ```suggestion
Consider the results from running the naïve implementation.
```

| Edge     | Count |
|----------|-------|
| [1->2]   | 1     |
| [3->4]   | 1     |
| [4->5]   | 5     |
| [5->6]   | 5     |
| [6->4]   | 5     |
| [7->9]   | 1     |
| [9->10]  | 1     |

> We confirm that our results for optimal placement match the results from naive placement. In order to verify the remaining edges, we run the extrapolation algorithm discussed above by hand. The next step is to implement the extrapolation and automatically verify that the optimal placement pass produces the same results as the naive pass for all edges.

Reviewer: ```suggestion
We confirm that our results for optimal placement match the results from naïve placement. In order to verify the remaining edges, we run the extrapolation algorithm discussed above by hand. The next step is to implement the extrapolation and automatically verify that the optimal placement pass produces the same results as the naïve pass for all edges.
```
